gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java

```diff
@@ -301,30 +301,42 @@ public static SetVariablesResponse toSetVariablesResponse(
    */
   public static JobActivationResult toActivateJobsResponse(
       final long key, final JobBatchRecord brokerResponse, final long maxResponseSize) {
-    final ActivateJobsResponse.Builder responseBuilder = ActivateJobsResponse.newBuilder();
-
     final Iterator<LongValue> jobKeys = brokerResponse.jobKeys().iterator();
     final Iterator<JobRecord> jobs = brokerResponse.jobs().iterator();
 
     long currentResponseSize = 0L;
     final List<ActivatedJob> sizeExceedingJobs = new ArrayList<>();
+    final List<ActivatedJob> responseJobs = new ArrayList<>();
 
     while (jobKeys.hasNext() && jobs.hasNext()) {
       final LongValue jobKey = jobKeys.next();
       final JobRecord job = jobs.next();
       final ActivatedJob activatedJob = toActivatedJob(jobKey.getValue(), job);
 
-      // +2 is for the metadata size that is added to ActivateJobsResponse when a new ActivatedJob
-      // is added to it
-      currentResponseSize += activatedJob.getSerializedSize() + 2;
-      if (currentResponseSize <= maxResponseSize) {
-        responseBuilder.addJobs(activatedJob);
+      final int activatedJobSize = activatedJob.getSerializedSize();
+      if (currentResponseSize + activatedJobSize <= maxResponseSize) {
+        responseJobs.add(activatedJob);
+        currentResponseSize += activatedJobSize;
       } else {
         sizeExceedingJobs.add(activatedJob);
       }
     }
 
-    return new JobActivationResult(responseBuilder.build(), sizeExceedingJobs);
+    ActivateJobsResponse response =
+        ActivateJobsResponse.newBuilder().addAllJobs(responseJobs).build();
+    // Response size can still exceed the maximum response size because of the metadata added on
+    // building the response. Therefore, we check the response size again and if the response size
+    // is still exceeding the maximum response size, we remove the last added job from the response
+    // and add it to the list of jobs to be reactivated.
+    // We do this until the response size is below the maximum response size.
+    if (response.getSerializedSize() > maxResponseSize) {
+      while (!responseJobs.isEmpty() && response.getSerializedSize() > maxResponseSize) {
```

Comment:
🔧 You could remove the if statement, as the while already covers it on its own.

Reply:
Ahh, right, I missed it after the refactor. 👍